Visitor Can Register For An Account

FlisAnn opened the pull request:
[ Visitors can Register for an account  to buy a subscription-Client](https://www.pivotaltracker.com/story/show/176055656)
### User Story
```
As a Visitor,
In order to access unique features on the news app,
I would like to be able to register for an account on the site.
```
## Changes proposed in this pull request:
* Adds a nested modal to include registration and subscription forms
* Adds functionality to registration portion of form
* Refactors `rootReducer` and API request functionality
* Updates UI styling for header menu
## What I have learned working on this feature:
* To make use of semantic UIs nested modal
* Linking `Form` submission functionality from separate component using `props`
* Combining functionality into a single `onClick` / `onSubmit` event
## Packages/Gems added
*`j-tockauth`
## Screenshots (Client)
![Screenshot 2021-01-14 at 14 29 43](https://user-images.githubusercontent.com/30779084/104597038-1c3f3980-5675-11eb-8f65-84e33182c975.png)
![Screenshot 2021-01-14 at 14 29 30](https://user-images.githubusercontent.com/30779084/104597077-295c2880-5675-11eb-8e1e-55e82ff38a15.png)
![Screenshot 2021-01-14 at 14 29 13](https://user-images.githubusercontent.com/30779084/104597127-37aa4480-5675-11eb-913d-54499b33989e.png)

FlisAnn committed:
Adding test

swardbladpaulo committed:
modify visitorCanRegisterForAnAccount test

swardbladpaulo committed:
add registration button to app, format code

swardbladpaulo committed:
add visitor can register fixture file

swardbladpaulo committed:
add initial sad path test for registration

swardbladpaulo committed:
fix the error we where reciving when the test was xhr get request error on the successful test

swardbladpaulo committed:
formatting

FlisAnn committed:
Updating test case. Adding new file for handling authentication. Connecting file to App component. Installing Jtock-Auth

FlisAnn committed:
Setting new case for SET_CURRENT_USER in rootReducer

FlisAnn committed:
fixing typos in various files

FlisAnn committed:
removed import of react

FlisAnn committed:
Cleaning up files and making test go green expect for not finding header-user-mail, which is expected

FlisAnn committed:
Changing payload response

FlisAnn committed:
Removing curlies that were in the wrong place, causing import not to work

FlisAnn committed:
putting back error.log

rithreaksa committed:
WIP for registration feature

FlisAnn committed:
changing auth to authenticatedUser
adding funcion for calling performAuthentication
setting authenticatedUser in initialState to false

swardbladpaulo committed:
debugging, modify fixture visitor can register, modify test for user can register, modify header.jsx, modify registrationform.jsx, modify auth payload for type current user to auth: true, got response from debugger

swardbladpaulo committed:
correct typo for visitor can register fixture, modify test visitorCanREgisterForAnAccount.feature.js, move authenticatedUser display message to registerForm and also modify it, format code in registrationForm, modify the response for the register user, modify the rootReducers, comment back in code in initialState

FlisAnn committed:
Fixing styling on request from PO, first removing updated-info, and then removing content from article card so it does not shows on index

FlisAnn committed:
Adding logo, adding menu, moving register button to menu

FlisAnn committed:
Adding styling, color to buttons

sealfury committed:
Creates header menu with logo

sealfury committed:
Creates nested modal
Adds onclick functionality to nested modal
moves registration functionality to nested modal

sealfury committed:
Adjusting error messages and test

emtalen commented on src/modules/auth.js:
```diff
@@ -0,0 +1,37 @@
+        payload: "Registration unsuccessful",
+      });
+      localStorage.removeItem("J-tockAuth-Storage");
+    }
```
there is no need for this if statement, your default role in the API is registered_user so the role will never be anything different when you sign up

sealfury committed:
Adjusts naming conventions, adds authenticated status to root reducer, adjusts logic in authorization, adjusts modal action logic accordingly

emtalen reviewed (commented):
I can't see where you are using the `auth` state and therefor it is difficult to see what is not working. We can have a look after the staff meeting

FlisAnn committed:
Accepting incoming change from Williams version.

sealfury committed:
adjusting test to proceed through registration form properly

sealfury committed:
Adjusting cypres test so that test clicks on submit button in modal
Refactoring modules and components
Changing error dispatch in article show function

FlisAnn committed:
Accepting incoming changes

FlisAnn commented on src/components/MasterModal.jsx:
```diff
@@ -0,0 +1,84 @@
+  };
+
+  const finalizePayment = () => {
+    // Stripe payment functionality will be added here
```
Does this comment need to be here?

FlisAnn commented on src/components/MasterModal.jsx:
```diff
@@ -0,0 +1,84 @@
+      >
+        <Modal.Header>Enter Your Details</Modal.Header>
+        <Modal.Content>
+          {/* <RegistrationForm /> */}
```
This is commented out?

sealfury replied:
Removed the commented out code in both instances

FlisAnn reviewed (commented):
Just some commented out code left, otherwise it looks good!

FlisAnn committed:
Merge branch 'development' into visitor_can_register_for_an_account

sealfury committed:
Removing comments from Register Modal

swardbladpaulo committed:
Merging cherry-picked changes

sealfury committed:
adds api routing conditional to index, removes tests for article index content display

emtalen commented on cypress/integration/visitorCanRegisterForAnAccount.feature.js:
```diff
@@ -0,0 +1,71 @@
+        url: "http://localhost:3000/api/**",
+        status: "401",
+        response: {
+          errors: "Registration unsuccessful",
```
Look at devise token auth's documentation and update this error message and the way you receive it

sealfury replied:
Updated

emtalen commented on src/components/DisplayArticlesData.jsx:
```diff
@@ -21,7 +21,8 @@ const DisplayArticlesData = () => {
 
   return (
-    <>{mainArticles.length > 0 && <div data-cy="index">{articleIndex}</div>}</>
+    
```
remove this extra space here

emtalen commented on src/components/MasterModal.jsx:
```diff
@@ -0,0 +1,82 @@
+          <Modal.Content>Card payment form goes here</Modal.Content>
+          <Modal.Description></Modal.Description>
+          <Modal.Actions>
+            <Button icon="check" content="All Done" onClick={finalizePayment} />
```
I would argue this is a liiiittle bit outside of the scope for this feature

sealfury replied:
Agreed, could have just as well called that `{closeBoth}` for now but I figured it served as a good indicator of how to proceed with implementation on the payment feature

emtalen commented on src/components/MenuHeader.jsx:
```diff
@@ -0,0 +1,30 @@
+      </Menu>
+    </Segment>
+  )
+
```
remove extra line

emtalen commented on src/index.js:
```diff
@@ -8,17 +8,20 @@ import { Provider } from 'react-redux';
+axios.defaults.baseURL =
+  process.env.NODE_ENV === "production"
+    ? "https://dream-time-news-api.herokuapp.com"
+    : "http://localhost:3000/api";
```
please go back to use an .env file

emtalen commented on cypress/fixtures/visitor_can_register.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "success": true,
```
please double check that this is really part of the response and received as a boolean

sealfury replied:
Have added logic to convert the response to boolean and have updated the fixture to reflect the actual response

sealfury committed:
editing test/fixture to match d-t-auth
formatting and adding logic to translate j-tockauth json results to boolean for authenticated state

tochman commented on src/components/MenuHeader.jsx:
```diff
@@ -0,0 +1,29 @@
+        <Image 
+        src="./assets/dreamtimesnews.png"
+        size="large"
+        fluid="true" />
```
Please fix code formatting

tochman commented on src/components/SingleArticle.jsx:
```diff
@@ -21,19 +21,18 @@ const SingleArticle = () => {
       </Button>
       {errorMessage && (
-        <Container data-cy="error-message-article">
+        <Message color="red" data-cy="error-message-article">
```
I would suggest 

```
  <Message
    header={errorMessage}
    data-cy="error-message-article"
  />
  ```
No need to pass in any html

sealfury committed:
fixing formatting and removing superfluous html from single article component

tochman reviewed (approved):
Okay, finally...
